File: Problem-lists/Algorithm_Patterns/Array/56_Merge_Intervals.js
/*
    Time: O(nlogn), space: O(n)
*/

/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
    const ans = [];

    // sort by start time
    intervals.sort((a, b) => a[0] - b[0]);

    for (const [start, end] of intervals) {
        if (ans.length > 0 && ans[ans.length - 1][1] >= start) {
            ans[ans.length - 1][1] = Math.max(ans[ans.length - 1][1], end);
        } else {
            ans.push([start, end]);
        }
    }

    return ans;
};

File: Problem-lists/Algorithm_Patterns/Array/56_Merge_Intervals.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

/*
    Given an array of intervals where intervals[i] = [starti, endi]
    Merge all overlapping intervals, and return an array of the non-overlapping intervals that cover all the intervals in the input.
*/

vector<vector<int>> merge(vector<vector<int>>& intervals) {
    vector<vector<int>> ans;

    sort(intervals.begin(), intervals.end());

    ans.push_back(intervals[0]);
    for (int i = 1; i < intervals.size(); i++) {
        if (intervals[i][0] <= ans[ans.size() - 1][1]) {
            ans[ans.size() - 1][1] = max(ans[ans.size() - 1][1], intervals[i][1]);
        } else {
            ans.push_back(intervals[i]);
        }
    }

    return ans;
}